tasks/eva_2406/gather_release_species.py

```diff
@@ -0,0 +1,357 @@
+#!/usr/bin/env python
+import csv
+import json
+import operator
+import os
+import re
+from argparse import ArgumentParser
+from collections import defaultdict
+
+import atexit
+
+import pandas as pd
+import psycopg2
+import psycopg2.extras
+import requests
+from ebi_eva_common_pyutils.config_utils import get_pg_metadata_uri_for_eva_profile
+from ebi_eva_common_pyutils.logger import logging_config
+from ebi_eva_common_pyutils.pg_utils import execute_query, get_all_results_for_query
+
+logger = logging_config.get_logger(__name__)
+logging_config.add_stdout_handler()
+
+
+eutils_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/'
+esearch_url = eutils_url + 'esearch.fcgi'
+esummary_url = eutils_url + 'esummary.fcgi'
+efetch_url = eutils_url + 'efetch.fcgi'
+ensembl_url = 'http://rest.ensembl.org/info/assembly'
+
+
+cache_file = 'cache.json'
+
+
+def load_cache():
+    global cache
+    if os.path.exists(cache_file):
+        with open(cache_file) as open_file:
+            cache.update(json.load(open_file))
+
+
+def save_cache():
+    with open(cache_file, 'w') as open_file:
+        json.dump(cache, open_file)
+
+
+atexit.register(save_cache)
+
+
+cache = defaultdict(dict)
+load_cache()
+
+
+def retrieve_assembly_summary_from_species_name(species):
+    """Search for all ids of assemblies associated with a species by depaginating the results of the search query"""
+    payload = {'db': 'Assembly', 'term': '"{}[ORGN]"'.format(species), 'retmode': 'JSON', 'retmax': 100}
+    response = requests.get(esearch_url, params=payload)
+    data = response.json()
+    search_results = data.get('esearchresult', {})
+    id_list = search_results.get('idlist', [])
+    while int(search_results.get('retstart')) + int(search_results.get('retmax')) < int(search_results.get('count')):
+        payload['retstart'] = int(search_results.get('retstart')) + int(search_results.get('retmax'))
+        response = requests.get(esearch_url, params=payload)
+        data = response.json()
+        search_results = data.get('esearchresult', {})
+        id_list += search_results.get('idlist', [])
+    response = requests.get(esummary_url, params={'db': 'Assembly', 'id': ','.join(id_list), 'retmode': 'JSON'})
+    summary_list = response.json()
+    if summary_list and 'result' in summary_list:
+        return [summary_list.get('result').get(uid) for uid in summary_list.get('result').get('uids')]
+
+
+def most_recent_assembly(assembly_list):
+    """Based on assembly summaries find the one submitted the most recently"""
+    if assembly_list:
+        return sorted(assembly_list, key=operator.itemgetter('submissiondate'))[-1]
```

Comment:
This definitely does make sense. But I am thinking if we should check with NCBI for pitfalls with this method (for ex: are updates made to older assemblies captured in this date? etc.,)

Reply:
Your concern is that this might not return the best (most relevant) assembly. 
I agree which is why there is another function to return the most complete assembly.